# **Sprint 1 Mini Project**


## Sprint 학습 목표

- EDA & Data Wrangling 과정을 코드로 구현할 수 있다.
- 목적에 맞는 시각화를 활용하여 비지니스 인사이트를 도출할 수 있다.

## Project 문제 정의

### 배경
> 여러분은 웹사이트와 더불어 오프라인 매장을 함께 운영하고 있는 **자전거 거래 플랫폼 회사**에서 데이터 분석 업무를 담당하고 있습니다.<br>
해당 회사의 주요 타겟 지역은 **호주**이며, 매출액을 늘리기 위해서는 자전거 거래 데이터를 분석하여 **인사이트를 도출**하는 작업이 필요합니다.<br>
회사 Database에 적재 되어있는 데이터를 추출하여 확보한 **고객 데이터**와 **거래 데이터**를 분석에 용이한 형태로 정제하는 것이 이번 프로젝트의 주요 과제입니다.<br>
가능하다면 데이터 분석을 통해 현재 상황과 매출액 증가를 위한 방안에 대한 **비즈니스 인사이트를 도출**하고 **실행 가능한 계획까지 수립**하는 프로젝트를 완성해보세요.




### 목표

#### 기본 과제
> - 주어진 데이터셋을 **이해하고 분석**할 수 있다.
- 주어진 질문에 대해 **시각화를 통해 설명**할 수 있다.
- **품질의 문제와 구조적 문제**를 파악하여 **정제**할 수 있다.


#### 도전 과제
> - 현재 플랫폼 내 자전거 거래의 **동향(trend)**을 파악한다.<br>
- **고객 세분화(Customer Segmentation)**를 통해 비슷한 특성을 가진 **고객군 각각의 특징을 파악**한다. <br>
- 고객군 별로 매출을 증가시키기 위해 적용할 수 있는 마케팅 전략을 수립하고 **실행 계획(action plan)**을 제시할 수 있다.

## **N111**

### 오늘의 목표
- 주어진 데이터셋을 불러올 수 있다.
- 데이터셋에 대해 pandas 메서드를 통해 기본 정보를 확인할 수 있다.


### `Topic 1`: 데이터셋 불러오기
본격적인 분석에 앞서, 아래의 데이터셋 설명을 통해 어떤 정보를 가진 데이터셋인지 파악 후 코드를 통해 데이터셋을 불러오도록 합시다.

<br>

> **customer 데이터셋 설명 (description)** <br>
우리 플랫폼에 등록된 고객 정보 데이터

- `customer_id`: 고객 고유번호
- `first_name`: 이름
- `last_name`: 성
- `gender`: 성별
- `past_3_years_bike_related_purchases`: 자전거 관련 구매 이력 (3년)
- `DOB`: 생년월일
- `job_title`: 직업명
- `job_industry_category`: 직무
- `wealth_segment`: 재력
- `deceased_indicator`: 사망 여부
- `default`: 정보
- `owns_car`: 차 소유 여부
- `tenure`: 재직 기간


#### **1-1. 라이브러리와 메서드는 무엇인가요?**
> 앞으로 여러분은 다양한 라이브러리를 사용하여 분석, 모델링을 진행하시게 될 텐데요.
>
> 오늘 Lecture Note에서도 pandas라는 라이브러리를 활용하고 있습니다.
>
> 그렇다면 라이브러리는 무엇이며, 왜 사용을 하는 것일까요? 이와 더불어 라이브러리와 함께 자주 등장하는 개념인 메서드는 무엇인지에 대해서도 함께 찾아본 후 내용을 정리하여 노션의 **진행내용**에 남겨주세요.


#### **1-2. pandas의 `read_csv` 함수를 활용하여 주어진 csv 파일(customer_info.csv)을 `customer`라는 이름으로 불러오세요.**


In [5]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [7]:
df = pd.read_csv("/Users/bumsoojoe/Desktop/Study/Bootcamp_Review/Section1/Mini_project/csv/customer_info.csv")

In [8]:
df.head(5)

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases         DOB               job_title  \
0                                   93  1953-10-12     Executive Secretary   
1                                   81  1980-12-16  Administrative Officer   
2                                   61  1954-01-20      Recruiting Manager   
3                                   33  1961-10-03                     NaN   
4                                   56  1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  \
0                                                 "'      Yes    11.0   
1                       <script>alert('hi')</script>      Yes    16.0   
2                                             01-Feb      Yes    15.0   
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0   
4                                                NIL      Yes     8.0   

   Unnamed: 13  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN

In [9]:
df.tail(5)

customer_id first_name last_name  gender  \
3995         3996    Rosalia  Halgarth  Female   
3996         3997     Blanch    Nisuis  Female   
3997         3998     Sarene   Woolley       U   
3998         3999  Patrizius       NaN    Male   
3999         4000      Kippy   Oldland    Male   

      past_3_years_bike_related_purchases         DOB              job_title  \
3995                                    8  1975-08-09  VP Product Management   
3996                                   87  2001-07-13        Statistician II   
3997                                   60         NaN      Assistant Manager   
3998                                   11  1973-10-24                    NaN   
3999                                   76  1991-11-05   Software Engineer IV   

     job_industry_category     wealth_segment deceased_indicator    default  \
3995                Health      Mass Customer                  N  -1.00E+02   
3996         Manufacturing     High Net Worth                  N       test   
3997                    IT     High Net Worth                  N        NaN   
3998         Manufacturing  Affluent Customer                  N        NaN   
3999                   NaN  Affluent Customer                  N        0/0   

     owns_car  tenure  Unnamed: 13  
3995       No    19.0          NaN  
3996      Yes     1.0          NaN  
3997       No     NaN          NaN  
3998      Yes    10.0          NaN  
3999       No    11.0          NaN

### `Topic 2`: 기본 정보 확인
> 위에서 불러온 데이터셋에 대해 pandas 메서드를 활용하여 기본 정보를 확인해봅시다.

#### **2-1. `customer`는 몇 개의 행과 열로 구성되어 있는지 확인해보세요.**

> 💡 **Hint)** shape

In [11]:
df.shape

(4000, 14)

#### **2-2. `customer`의 정보(column 이름, 결측치 개수, 데이터타입 ...)를 pandas 메서드를 활용하여 확인해보세요.**
> 💡 **Hint)** info

- 몇 개의 column을 가지고 있는 데이터셋인지 확인해보세요.
  - 각 column에 어떤 정보가 들어있는지 확인한 뒤, 무의미한 데이터가 들어있는 column이 있는지도 함께 논의해보세요.

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 14 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   customer_id                          4000 non-null   int64  
 1   first_name                           4000 non-null   object 
 2   last_name                            3875 non-null   object 
 3   gender                               4000 non-null   object 
 4   past_3_years_bike_related_purchases  4000 non-null   int64  
 5   DOB                                  3913 non-null   object 
 6   job_title                            3494 non-null   object 
 7   job_industry_category                3344 non-null   object 
 8   wealth_segment                       4000 non-null   object 
 9   deceased_indicator                   4000 non-null   object 
 10  default                              2571 non-null   object 
 11  owns_car                      

In [42]:
dfColDtype = list(df.dtypes)
dfColDtype

[dtype('int64'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('int64'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('float64'),
 dtype('float64')]

#### **2-3. `customer`에 결측치가 있는지 확인한 뒤, 어떻게 처리하는 것이 좋을지 생각해보세요.**
> 💡 **Hint)** missing value, nan, isnull
- 결측치 처리 방식에는 크게 두 가지가 있습니다.
  - 결측치 대체: 특정 값으로 결측치를 채우는 것
  - 결측치 제거: 결측치가 있는 행 또는 열을 제거하는 것

- 결측치가 있는 column에는 어떤 것이 있는지 확인해보세요.
- 각 column 별로 결측치를 어떻게 처리하는 것이 좋을지 생각해보세요.
  - 결측치 처리에 무조건적인 정답은 없으며, 조에서 논의 후 내린 결론을 논리적으로 설명하는 것이 무엇보다 중요합니다.

In [16]:
df.isnull().sum()

customer_id                               0
first_name                                0
last_name                               125
gender                                    0
past_3_years_bike_related_purchases       0
DOB                                      87
job_title                               506
job_industry_category                   656
wealth_segment                            0
deceased_indicator                        0
default                                1429
owns_car                                  0
tenure                                   87
Unnamed: 13                            4000
dtype: int64

#### **2-4. `customer`에 중복값이 있는지 확인한 뒤, 있다면 제거하세요.**
> 💡 **Hint)** duplicated

In [19]:
df.duplicated().sum()

0

In [24]:
colList = df.columns
for i in colList:
    print(i," = ", df[i].duplicated().sum())

customer_id  =  0
first_name  =  861
last_name  =  274
gender  =  3994
past_3_years_bike_related_purchases  =  3900
DOB  =  551
job_title  =  3804
job_industry_category  =  3990
wealth_segment  =  3997
deceased_indicator  =  3998
default  =  3932
owns_car  =  3998
tenure  =  3977
Unnamed: 13  =  3999


In [34]:
len(colList)

14

In [43]:
for i in range(len(colList)):
    print("- ",colList[i], " : ", dfColDtype[i])

-  customer_id  :  int64
-  first_name  :  object
-  last_name  :  object
-  gender  :  object
-  past_3_years_bike_related_purchases  :  int64
-  DOB  :  object
-  job_title  :  object
-  job_industry_category  :  object
-  wealth_segment  :  object
-  deceased_indicator  :  object
-  default  :  object
-  owns_car  :  object
-  tenure  :  float64
-  Unnamed: 13  :  float64


### **Conclusion**
> 오늘 논의한 내용을 바탕으로 아래 항목에 대해 논의하고 답변을 정리해보세요.

 * 해당 데이터셋은 어떻게 구성되어 있는지 정리해보세요.
  - 몇 개의 행과 열로 구성되어 있는 데이터셋인가요?
    - 행 : 4000
    - 열 : 14
  - 각 컬럼이 어떤 데이터 타입에 속해있나요?
    -  customer_id  :  int64
    -  first_name  :  object
    -  last_name  :  object
    -  gender  :  object
    -  past_3_years_bike_related_purchases  :  int64
    -  DOB  :  object
    -  job_title  :  object
    -  job_industry_category  :  object
    -  wealth_segment  :  object
    -  deceased_indicator  :  object
    -  default  :  object
    -  owns_car  :  object
    -  tenure  :  float64
    -  Unnamed: 13  :  float64
 * 데이터 전처리 과정을 정리해보세요. (결측치 처리, 중복값 제거)
  - 결측치는 몇 개가 있었나요? 어떤 기준으로 제거/대치하셨나요?
    - customer_id                               0
    - first_name                                0
    - last_name                               125
    - gender                                    0
    - past_3_years_bike_related_purchases       0
    - DOB                                      87
    - job_title                               506
    - job_industry_category                   656
    - wealth_segment                            0
    - deceased_indicator                        0
    - default                                1429
    - owns_car                                  0
    - tenure                                   87
    - Unnamed: 13                            4000
  - 중복값은 몇 개가 있었나요? 어떤 메소드로 제거하셨나요?
    - 중복값 : 0
    - 중복되는 열은 없다.


In [44]:
# 결측치 전처리
"""
결측치가 있는 컬럼을 직접적으로 확인 너무 많은 결측치가 있는 행이나 열은 제거한다.

"""
df

customer_id      first_name  last_name  gender  \
0               1         Laraine  Medendorp       F   
1               2             Eli    Bockman    Male   
2               3           Arlin     Dearle    Male   
3               4          Talbot        NaN    Male   
4               5  Sheila-kathryn     Calton  Female   
...           ...             ...        ...     ...   
3995         3996         Rosalia   Halgarth  Female   
3996         3997          Blanch     Nisuis  Female   
3997         3998          Sarene    Woolley       U   
3998         3999       Patrizius        NaN    Male   
3999         4000           Kippy    Oldland    Male   

      past_3_years_bike_related_purchases         DOB               job_title  \
0                                      93  1953-10-12     Executive Secretary   
1                                      81  1980-12-16  Administrative Officer   
2                                      61  1954-01-20      Recruiting Manager   
3                                      33  1961-10-03                     NaN   
4                                      56  1977-05-13           Senior Editor   
...                                   ...         ...                     ...   
3995                                    8  1975-08-09   VP Product Management   
3996                                   87  2001-07-13         Statistician II   
3997                                   60         NaN       Assistant Manager   
3998                                   11  1973-10-24                     NaN   
3999                                   76  1991-11-05    Software Engineer IV   

     job_industry_category     wealth_segment deceased_indicator  \
0                   Health      Mass Customer                  N   
1       Financial Services      Mass Customer                  N   
2                 Property      Mass Customer                  N   
3                       IT      Mass Customer                  N   
4                      NaN  Affluent Customer                  N   
...                    ...                ...                ...   
3995                Health      Mass Customer                  N   
3996         Manufacturing     High Net Worth                  N   
3997                    IT     High Net Worth                  N   
3998         Manufacturing  Affluent Customer                  N   
3999                   NaN  Affluent Customer                  N   

                                                default owns_car  tenure  \
0                                                    "'      Yes    11.0   
1                          <script>alert('hi')</script>      Yes    16.0   
2                                                01-Feb      Yes    15.0   
3     () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0   
4                                                   NIL      Yes     8.0   
...                                                 ...      ...     ...   
3995                                          -1.00E+02       No    19.0   
3996                                               test      Yes     1.0   
3997                                                NaN       No     NaN   
3998                                                NaN      Yes    10.0   
3999                                                0/0       No    11.0   

      Unnamed: 13  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  
...           ...  
3995          NaN  
3996          NaN  
3997          NaN  
3998          NaN  
3999          NaN  

[4000 rows x 14 columns]

In [45]:
df.gender.value_counts()

gender
Female    2037
Male      1872
U           88
F            1
Femal        1
M            1
Name: count, dtype: int64

In [47]:
df.job_title.value_counts()

job_title
Business Systems Development Analyst    45
Tax Accountant                          44
Social Worker                           44
Internal Auditor                        42
Recruiting Manager                      41
                                        ..
Database Administrator I                 4
Health Coach I                           3
Health Coach III                         3
Research Assistant III                   3
Developer I                              1
Name: count, Length: 195, dtype: int64